### PR TITLE
Sk Adds Config Handler Class and an Example Pipeline Workflow

### PR DESCRIPTION
**Pull request by scarlett-k-nhs**
# Summary

**What/Why**

This branch adds multiple functionality:

1. Functions to define prompt templates and save this to a defined path to then be able to read the data in using the experimental pipeline handler.
2. Yaml files, which are parsed and handled in the ExperimentConfigHandler Class.
4. An example experimental workflow where ollama, local, and gliner are compared.

# Checklists

This pull/merge request meets the following requirements:

- [ ] code runs
- [ ] [developments are ethical][data-ethics-framework]
- [ ] developments are secure (no data or notebook output)
- [ ] you have made proportionate checks that the code works correctly
- [ ] [minimum usable documentation][agilemodeling] written in the `docs` folder

Comments have been added below around the incomplete checks.

[agilemodeling]: http://agilemodeling.com/essays/documentLate.htm
[data-ethics-framework]: https://www.gov.uk/government/publications/data-ethics-framework

### CODE REVIEW

File: notebooks/creating_prompt_templates.ipynb [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+    "generate_template = \"\"\"[INST]\n",
+    "<<SYS>>\n",
+    "You are a medical student answering an exam question about writing clinical notes for patients.\n",
+    "<</SYS>>\n",
+    "\n",
+    "Keep in mind that your answer will be asssessed based on incorporating all the provided information and the quality of prose.\n",
+    "\n",
+    "1. Use prose to write an example clinical note for this patient's doctor.\n",
+    "2. Use less than three sentences.\n",
+    "3. Do not provide a recommendations.\n",
```
Comment (willpoulett) [inline notebooks/creating_prompt_templates.ipynb:80]: tiny typo 'a reccomendations'

Comment (willpoulett) [reply notebooks/creating_prompt_templates.ipynb:80]: also in experimental pipeline notebook

File: notebooks/experimental_pipeline_example.ipynb [hunk 2]
```diff
@@ -0,0 +1,1444 @@
...
+    "default_config_path = \"../config/experimental_config.yaml\""
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Then you need to define you overrides:\n",
+    "\n",
+    "* **iter_overrides**: For components you want to override that are iterated across. Meaning you want each value to be compared to every other value that is defined.\n",
```
Comment (willpoulett) [inline notebooks/experimental_pipeline_example.ipynb:184]: Can this sentence be a bit clearer? I am a bit confused by it

File: requirements.txt [hunk 3]
```diff
@@ -12,4 +12,12 @@ julia==0.6.2
...
-gliner==0.1.12
+gliner>=0.1.12
+hydra-core==1.3.2
+omegaconf==2.3.0
+pydantic
+graphviz
+mplcursors
+networkx
+plotly
+omegaconf
```
Comment (willpoulett) [inline requirements.txt:23]: do we want specific versions?

File: config/experimental_config.yaml [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+synthea:
+  population_num: '50'
+  county: 'West Yorkshire'
+
+generate:
+  llm_model_features:
+    llm_model_name: llama2
+    prompt_template_path: llama2_template.json
+
+extraction:
```
Comment (danjscho) [inline config/experimental_config.yaml:15]: Noting we have "structure" in some places, and "extraction" in others - should discuss and stick to one.

File: config/global_config.yaml [hunk 5]
```diff
@@ -0,0 +1,12 @@
+
+output_paths:
+  output_folder: ../outputs
+  extraction_template: ../src/config/templates/extraction/
```
Comment (danjscho) [inline config/global_config.yaml:4]: Is it worth having these templates at the root level, rather than in src?

File: src/old_config.py
Comment (danjscho) [inline src/old_config.py:1]: Do we need this or can it be aligned to the new setup?

File: outputs/experiment_0001_14_05_24/experimental_config.yaml
Comment (danjscho) [inline outputs/experiment_0001_14_05_24/experimental_config.yaml:1]: I suggest we move this to an `example_experiment...` sort of folder and .gitignore contents of the outputs folder (except the explicit example or default bits we want to keep)?

File: outputs/experiment_0001_14_05_24/experimental_config.yaml [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  local_features:
+    hf_repo_id: yuuko-eth/UniNER-7B-all-GGUF
+    hf_filename: UniversalNER-7B-all-Q4_K_M.gguf
+    prompt_template_path: universal_ner_template.json
+  ollama_features:
+    ollama_ner_model: zeffmuks/universal-ner
+    prompt_template_path: universal_ner_template.json
+  entity_list:
+  - person
+  - date of bith
```
Comment (danjscho) [inline outputs/experiment_0001_14_05_24/experimental_config.yaml:27]: typo

File: notebooks/full_pipeline_example.ipynb [hunk 7]
```diff
@@ -21,7 +21,7 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "from src.config import (\n",
+    "from src.old_config import (\n",
```
Comment (danjscho) [inline notebooks/full_pipeline_example.ipynb:24]: Old config vs config?

**Review by danjscho — COMMENTED** [review]
From a first pass to discuss this afternoon - will go back over once checked in.

File: notebooks/experimental_pipeline_example.ipynb [hunk 8]
```diff
@@ -0,0 +1,1444 @@
...
+    "\n",
+    "You will **NEED** to define your experiment name as this defines where your experiment folder should sit.\n",
+    "\n",
+    "* **combine_overrides**: Combine overrides is for users to define when they want to look at a combination of values, but not necesarrily iterate across all values. For example your specific llm model, might require a different prompt template.\n",
```
[the remark this answers is not in the document]
Comment (danjscho) [reply notebooks/experimental_pipeline_example.ipynb:188]: typo `necessarily`

**Review by danjscho — APPROVED** [review]
Looks good - will also add issue to highlight required fix around standardisation.

**Pull request merged**